python/ray/serve/tests/test_reconfiguration.py

```diff
@@ -0,0 +1,36 @@
+import pytest
+import asyncio
+from ray import serve
+import ray
+
+
+def test_reconfiguration():
+    @serve.deployment(max_concurrent_queries=10, num_replicas=3)
+    class A:
+        def __init__(self):
+            self.state = None
+
+        def reconfigure(self, config):
+            self.state = config
+
+        async def __call__(self):
+            # change sleep to SignalActor for testing
```

Comment:
We should avoid using `sleep` in our tests to prevent test flakiness.  Can you try using a `SignalActor` to signal when the call should return? https://docs.ray.io/en/latest/advanced.html#multi-node-synchronization-using-an-actor

Reply:
I don't know how to use a `SignalActor` in this situation, because `A.options(version="1", user_config={"a": 2}).deploy()` is a synchronous function.
Should I modify this test case to a replica layer test or keep using `asyncio.sleep`?

Reply:
I'm not sure how to use `SignalActor` in this situation, because `A.option().deploy() ` is a synchronous function, and I can't use a private function `_get_global_client()` and private variable `_controller` in the deployment to reconfigure asynchronously. So I have to use `SignalActor.send.remote()` before I reconfigure by `A.option().deploy()`.

I'm not sure if we should keep using `sleep` or just add a unit test which just creates a replica object, and test it's `handle_request` and `reconfigure` function. Maybe there's a better way to use `SignalActor`, what do you think? @architkulkarni

Reply:
I'm not sure how to use `SignalActor` in this situation, because `A.option().deploy() ` is a synchronous function, and I can't use a private function `_get_global_client()` and private variable `_controller` in the deployment to reconfigure asynchronously. So I have to use `SignalActor.send.remote()` before I reconfigure by `A.option().deploy()`.

I'm not sure if we should keep using `sleep` or just add a unit test which just creates a replica object, and test it's `handle_request` and `reconfigure` function. Maybe there's a better way to use `SignalActor`, what do you think? @architkulkarni

Reply:
Can you imitate the logic in `test_advanced.py::test_serve_graceful_shutdown`?  I didn't quite understand your concerns about reconfiguring asynchronously, why can't we just call `SignalActor.send.remote()` after the second call to `.deploy()`?  

I think it's the same logic as what you currently have, except you're replacing `await time.sleep(10)` with `await signalactor.wait.remote()`.  The old assumption was that the 10 second timer finishes after the second call to `.deploy()` (in other words, the reconfiguration happens during the `await time.sleep()`.). But in the revised version, the new guarantee is just that we call `SignalActor.send()` after the second call to `.deploy()`, so it's still true that the reconfiguration happens during an `await`.

Reply:
(Sorry I might have worded things confusingly, the reconfiguration doesn't actually happen during the `await` with this PR, but I just meant that's what would happen without the PR, and we're testing that that doesn't happen)

Reply:
If I'm understanding the code correctly, even though `deploy` is synchronous, it calls `replica_actor.reconfigure.remote()` which is asynchronous, so the reconfiguration can happen while the replica actor is in an `await`.

Reply:
I understand, and sorry to trouble you.

I updated again, maybe this version is what you want